Add Lecture 12 Notes

epuer18 opened the pull request:
Add Lecture_12.md to the repository. The file is divided into three main sections on Testing Review, Mutation Testing, and Test Input Generation. The earliest prefaces the lecture content with notes on testing metrics and a refresher on coverage. Mutation Testing proposes a new test suite metric and Input Generation provides a breakdown into the components of a test case. Provide example code blocks for mutation operators and branching coverage.

epuer18 committed:
Add Lecture 12 from Template

Add Lecture_12.md file to the repository with content from the template
proposed in PR #43. Update the date, lecture number, and topic in the
file header. The file will contain notes from Lecture 12 on mutation
testing and test input generation.

epuer18 committed:
Add Testing Review and Mutation Testing Section

Add the Testing Review section to Lecture_12.md containing a review on
test suite metrics, namely code coverage. Add Mutation Testing Section
divided into subsections about Bugs/Defects and Mutants. Provide a code
example on Mutation Operators in the Bugs/Defects subsection.

Update lecture content up to timestamp <1:04:56> of Panopto CS-340 Su21
Lecture 12 Video.

epuer18 committed:
Fix Example Code Block

The code block for example common mutant operators in Lecture_12.md was
faulty as it used (''') as opposed to (```). Fix this bug by using the
correct symbols. Remove an extra (*) from the Coupling Hypothesis
bolding.

epuer18 committed:
Add Input Generation and Finish Lecture 12 Notes

Finish the Mutation Testing section by adding a subsection on the
limitations of the method. Add a section on Input Generation with
subsections for test input and comparators. Add an example python code
block to explain branch and path coverage, this will serve as motivation
for theorem provers and automating test input. Remove remaining code
from the template.

epuer18 committed:
Fix Spelling Bugs in Lecture 12 Notes

Lecture_12.md contained spelling bugs across all sections. Utilize
Grammarly spellchecker to localize and fix spelling errors.

Examples of Defect Corrections:

- "statment" to "statement"
- "Rudamentary" to "Rudamentary"
- "abput" to "about"
- Add comma after "However"
- Combine "bloodhounds"
- "artifically" to "artificially"
- Remove "actually"

kevinaangstadt commented:
@AlphaCentauri79 could you please review this PR?  Be sure to look at both content as well as the commits and messages.  Because of permissions, you will likely need to complete your review in the comments here.

DigitalSRC reviewed (commented):
This looks pretty good. Im getting confused cause you organized the order from what it appears in our notes, while it is nice it was somewhat confusing, id suggest maybe adding a note to commits on this.

Bugs and Defects: (optional)
- For additional info, the bug was found in a vacuum tube machine on a boat. 

Mutants:
- The Coupling Hypothesis is more of a larger idea, in a bullet point system this should be the topic rather than additional information, even bold it was a bit hard to find.

Test Input Generation:
- spelling test input I am pretty sure is not solely dates, dates -> data
- Id honestly put the comparator model right after the definition, sense its mentioned its easier to define at this stage.

Didnt want to draw the CFG??? gotta step up your game. Im kidding. Anyways it all looks nice and neat. Think the few changes I mentioned should suffice, other than that its good to go.

epuer18 committed:
Specify Bug in Lecture 12

PR #46 raises how the first bug to a program in the Bugs and Defects of
Lecture_12.md was an insect in a vacuum tube on a boat. Update to
provide this more specific (and rather funnier) fact.

epuer18 committed:
Reorganize Sub-Sections in Input Gen for Lecture 12

The previous version of Lecture_12.md had the subsections in test input
generation follow the structure of the definition, namely Test Input
before Comparator. However, the chronological ordering of these sections
(which is the reverse) is more intuitive and portrays a more accurate
picture of the logic followed in class.

epuer18 committed:
Clarify Comparator Example Usage in Lecture 12

The Comparator subsection in Lecture_12.md provides two bullet points
with example cases of partial matches. Add a super-bullet point
clarifying that these are examples as provided in the lecture.

epuer18 committed:
Re-Format The Mutants Section in Lecture 12

Lecture_12.md contains a section on Mutation Testing where the Competent
Programmer Hypothesis (CPH) was an important point (hence in bold)
within the Mutants subsection. Update this to make the CPH its own
sub-section (triple hash ###). CPT is an important topic that also
contains significant content both explaining it and related to it.

Update from bullets to paragraphs. Bullet point content was written in
sentences, and the previous answer format was more fitting to bullets
than the section format, where prose fairs better.

epuer18 commented:
Awesome, thanks! Just added commits for all the things you mentioned which I agree with (I struggled to catch how the specific bug was in a vacuum tube in my rewatch of the lecture but thanks for the comment). 

For the 'dates' I was just providing an example of a partial match as opposed to a spelling mistake, so I just made a commit clarifying that by changing the phrasing a bit and adding a question before.
